### PR TITLE
Code-behind Animation Trigger

### PR DESCRIPTION
**Pull request by jmacato**
This template is not intended to be prescriptive, but to help us review pull requests it would be useful if you included as much of the following information as possible:

- What does the pull request do?
  Adds a Done event and a Run method for Animations, which is going to be utilized by code-behind animation scenarios.

### CODE REVIEW

File: src/Avalonia.Animation/Animation.cs [hunk 1]
```diff
@@ -153,5 +160,28 @@ public IDisposable Apply(Animatable control, IObservable<bool> matchObs)
...
+
+        /// <inheritdocs/>
+        public Task Run(Animatable control)
+        {
+            var tcs = new TaskCompletionSource<object>();
+
+            if (this.RepeatCount == RepeatCount.Loop)
+                tcs.SetException(new InvalidOperationException("Looping animations must not use the Run method."));
+
+            this.Done += delegate
```
Comment (jkoritzinsky) [inline src/Avalonia.Animation/Animation.cs]: We should check that the sender parameter is equal to control so we can run the same animation on multiple controls programmatically.

**Review by jkoritzinsky — CHANGES_REQUESTED** [review]

File: src/Avalonia.Animation/Animation.cs (hunk 1 above)
Comment (jmacato) [reply src/Avalonia.Animation/Animation.cs]: Good catch!

File: src/Avalonia.Animation/Animation.cs [hunk 2]
```diff
@@ -153,5 +160,31 @@ public IDisposable Apply(Animatable control, IObservable<bool> matchObs)
...
+
+            if (this.RepeatCount == RepeatCount.Loop)
+                run.SetException(new InvalidOperationException("Looping animations must not use the Run method."));
+
+            this.Done += (sender, args) =>
+            {
+                if (sender == control)
+                    run.SetResult(null);
+                else
+                    run.SetCanceled();
```
Comment (jkoritzinsky) [inline src/Avalonia.Animation/Animation.cs]: Why are we canceling the task here? This wouldn't allow users to run the same animation on two different controls at the same time.

Comment (jmacato) [reply src/Avalonia.Animation/Animation.cs]: I was thinking of making it cancel the task when an incongruity occurs but yeah, it's the wrong behaviour

File: src/Avalonia.Animation/AnimatorStateMachine`1.cs [hunk 3]
```diff
@@ -243,7 +243,10 @@ private void InternalStep(PlayState _playState, Func<double, T, T> Interpolator)
                         {
                             _targetControl.SetValue(_parent.Property, _lastInterpValue, BindingPriority.LocalValue);
                         }
+
                         _targetObserver.OnCompleted();
+                        _targetAnimation.SetDone(_targetControl);
```
Comment (jkoritzinsky) [inline src/Avalonia.Animation/AnimatorStateMachine`1.cs]: I've been doing some tests by re-implementing the page transitions, and I found a bug here. If the user animates a render transformation, then the value of `_targetControl` is the Transform object, not the control that is being transformed. This causes the animation to never be registered as completed.

**Review by jkoritzinsky — APPROVED** [review]

**Pull request merged**